File: Visual Basic.NET/Forms Applications/Demo/frmProgress.vb
Public Class frmProgress

  Private m_bStopped As Boolean = False

  Public Sub New(ByVal sTitle As String, ByVal nMax As Integer)
    MyBase.New
    InitializeComponent()
    Me.Text = sTitle
    lblAction.Text = sTitle
    lblDetails.Text = ""
    prgProgress.Maximum = nMax
  End Sub

  Public Function DoProgress(ByVal nCurrent As Integer, Optional ByVal sDetail As String = "") As Boolean
    If (nCurrent > prgProgress.Maximum) Then
      nCurrent = prgProgress.Maximum
    End If

    prgProgress.Value = nCurrent
    lblDetails.Text = sDetail
    Application.DoEvents()
    Return Not Me.m_bStopped
  End Function

  Private Sub btnStop_Click(ByVal sender As Object, ByVal e As EventArgs)
    Me.m_bStopped = True
  End Sub

End Class

File: Visual CSharp.NET/Forms Applications/Demo/frmProgress.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace Demo
{
  public partial class frmProgress : Form
  {
    private bool m_bStopped = false;

    public frmProgress(string sTitle, int nMax)
    {
      InitializeComponent();

      this.Text = sTitle;
      lblAction.Text = sTitle;
      lblDetails.Text = "";
      prgProgress.Maximum = nMax;
    }

    public bool DoProgress(int nCurrent, string sDetail = "")
    {
      if (nCurrent > prgProgress.Maximum)
        nCurrent = prgProgress.Maximum;
      prgProgress.Value = nCurrent;
      lblDetails.Text = sDetail;
      Application.DoEvents();
      return !m_bStopped;
    }

    private void btnStop_Click(object sender, EventArgs e)
    {
      m_bStopped = true;
    }
  }
}
